src/masoniteorm/models/Model.py

```diff
@@ -695,12 +695,31 @@ def first_or_create(cls, wheres, creates: dict = None):
             creates = {}
         self = cls()
         record = self.where(wheres).first()
+        if not record:
+            return self.create_or_first(wheres, creates)
+        return record
+
+    @classmethod
+    def create_or_first(cls, wheres, creates: dict = None):
+        """First attempts to create a record. If the record already exists then find the first record matching the
+        where clause.
+
+        Returns:
+            Model
+        """
+        if creates is None:
+            creates = {}
+        self = cls()
         total = {}
         total.update(creates)
         total.update(wheres)
-        if not record:
+
+        try:
             return self.create(total, id_key=cls.get_primary_key())
-        return record
+        except QueryException as e:
```

Comment:
i dont think we should be hijacking this exception its going to be thrown and overwrite legitimate  query exceptions right? like if an inmsert fails on the table?

also if thats not the case i think nyoure trying to check for the existence of a specific database error is thrown which you cant do because all diff databases and versions throw diff errors

Reply:
The solution to this would be to add a new exception. UniqueContraintViolationException to the connection classes. The different error codes for unique exception failures can be parsed on the connection class level. Allows for easier extensibility when adding support for new databases as well.

Reply:
i would be more comfortable with putting it in the connection class